### PR TITLE
Retries

### PR DESCRIPTION
**Pull request by jcorbin**
r @Raynos @kriskowal @rf

### CODE REVIEW

File: node/request.js [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+};
+
+TChannelRequest.prototype.shouldRetry = function shouldRetry(err, res, arg2, arg3) {
+    var self = this;
+
+    if (self.outReqs.length >= self.retryLimit) {
+        return false;
+    }
+
+    if (err) {
```
Comment (Raynos) [inline node/request.js]: can we check the `err.isErrorFrame` boolean ?

Comment (Raynos) [inline node/request.js]: If we do not check that boolean we will retry on all network errors; i.e. `ECONNRESET` and pals.

**Comment by kriskowal** [comment]
This looks solid to me. There are pending Travis errors, so I’m going to let this sit. We can cut a tchannel tag in the morning. I won’t have a PR based on this ready today regardless.

**Comment by Raynos** [comment]
Other then improving `shouldRetry()` with better error detection (having a default case where we always retry seems a bit loose); this looks good to me.

**Comment by jcorbin** [comment]
As confirmed with @mranney this morning, the entire point of the design here is to default into retrying with only blacklisted "why not".

This PR doesn't add any opt-out path, which should / will come as needed.  However we're intentionally going with the design of defaulted-in to retries.

Especially so for socket errors.

File: node/request.js [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+TChannelRequest.prototype.onReqDone = function onReqDone(err, res, arg2, arg3) {
+    var self = this;
+    var now = self.channel.timers.now();
+    self.elapsed = now - self.start;
+    if (self.elapsed < self.timeout &&
+        self.shouldRetry(err, res, arg2, arg3)) {
+        self._lastErr = err;
+        self._lastArg2 = arg2;
+        self._lastArg3 = arg3;
+        process.nextTick(deferResend);
```
Comment (Raynos) [inline node/request.js]: In the near future we will want some kind of backoff here.

**Comment by Raynos** [comment]
Are we sure all socket errors are remote errors ? Does the node `net` library emit `net` errors that are basically a `BadRequest`.

To be honest I don't mind either way as long as we have stats and information on what the hell is going through our processes.

File: node/request.js (hunk 2 above)
Comment (jcorbin) [inline node/request.js]: no we don't, backoff is a per-peer concern; there's no sane reason to blow your timeout deadline sleeping.

Comment (Raynos) [inline node/request.js]: true; makes sense.

**Pull request merged**